### ISSUE TITLE
Sagas performance issue

### ISSUE THREAD

**Issue by SeanFeldman** [Particular/NServiceBus.AzureStoragePersistence#23]
Currently Azure saga persister is not performant **at all**.
When saga is persisted, partition/row keys are assigned an internal GUID generated by NServiceBus which has nothing to do with the message data. When customers provide mappings to find saga instances, partition/row keys are not involved, causing azure storage full table scan. That's a performance killer. From [Azure documentation](https://azure.microsoft.com/en-us/documentation/articles/storage-table-design-guide/):

> As with keys in a relational database table, the PartitionKey and RowKey values are indexed to create a clustered index that enables fast look-ups; however, the Table service does not create any secondary indexes so these are the only two indexed properties.

The level of performance you are guaranteed is explicitly set at the partition/row key level. therefore, when you run a full table scan, its not efficient and doesn't have any guarantee of performance.

Performance (top to bottom) based on query types:

1. PartitionKey + RowKey
1. PartitionKey + RowKey + OtherProperty
1. PartitionKey + OtherProperty
1. OtherProperty <== **NSB today**

**Comment by Scooletz** [comment]
:+1: Going from full storage table scan to anything would be a great improvement! It reminded me good old http://www.troyhunt.com/2013/12/working-with-154-million-records-on.html

**Comment by DavidBoike** [comment]
Would a "pointer document" approach like Raven help? You'd have to then do 2 queries but both could be full matching Partition+Row keys.

**Comment by SeanFeldman** [comment]
@DavidBoike the issue with the "pointer document" approach (as we've discussed) is that rather than having 1 operation, now it's two. We could get away with this if we'd batch the 2 operations (insert records), but the issue is that both wouldn't need to have the same partition key. So we're back to square one.

**Comment by Scooletz** [comment]
@SeanFeldman Correct me if I'm wrong, but in my opinion @DavidBoike proposal make sense: we could insert N additional indexing entity which would consist of name of property + its value, pointing to the saga identifier. Having this, we'd pay the price once, when saga is instantiated (in v6 you can't change values of properties creating a natural key for a saga), but then, instead of one table scan we'd go with one key lookup by name+value of the property, which would return an entity containing only saga id. Then we'd query for the saga identifier. Instead of one full scan we have two lookups by partition + row key.
This would result in much lower IOPS overhead, lowering possibility of hitting the Azure account limitations.
Does it make sense?

**Comment by SeanFeldman** [comment]
@Scooletz what those N additional entities have as their PKey?

**Comment by Scooletz** [comment]
The `ISagaPersister` currently support following ways of querying for saga:
```
/// <summary>
/// Gets a sagaData entity from the persistence store by its Id.
/// </summary>
Task<TSagaData> Get<TSagaData>(Guid sagaId, SynchronizedStorageSession session, ContextBag context) 
where TSagaData : IContainSagaData;

/// <summary>
/// Looks up a sagaData entity by a given property.
/// </summary>
Task<TSagaData> Get<TSagaData>(string propertyName, object propertyValue, SynchronizedStorageSession session, ContextBag context) 
where TSagaData : IContainSagaData;
```

TL;DR:
My proposition could be summarized as building a secondary index in a key value store.

Details:
The first method getting saga is an obvious case. We can just get the saga by id. The second one uses the `propertyName` with a `propertyValue`. This can be combined by concatenation/serialization into one string. We can even hash it with SHA1. Having this, you can split the value in two (for SHA1 10 bytes for PartitionKey and 10 bytes for RowKey) and use as a PartitionKey/RowKey pair, and place `sagaId` as the value. To obtain the saga by the `propertyValue` first you do the look up by the pair (or its hash) to retrieve `sagaId`, and then do the lookup by identifier.
We save the scan, by introducing additional lookup by a key. For small number of Sagas this will add some overhead. For bigger sets, in my opinion, this can be a real improvement.

**Comment by SeanFeldman** [comment]
@Scooletz but when you write the entities, you can't have a different PKey. If you do, then your writes cannot be batched, and you will run into possibility of orphaned/partial entries.

**Comment by Scooletz** [comment]
@SeanFeldman Yes you can run into that situation, but if you order writer properly you can live with that. Consider the following sequence:
- generate saga identifier
- write secondary index entities pointing to nonexistent saga (by the generated id)
- write saga entity under the generated identifier

If the sequence breaks in any moment, you don't have the saga entity written and it means that either saga was deleted or created improperly. You'll try again to create the saga instance. Yes, it'll have a different id generated, but the values and keys for secondary indices will be the same. 
Does this reasoning make sense?

**Comment by SeanFeldman** [comment]
I see. So if saga data contains multiple properties, we're essentially going to create for each property an entity with a `propertyName` + `propertyValue` as a PKey, pointing to `sagaId`. And if we're failing to make the last write (sata entity), we'll have a bunch of orphaned entities.

2 more items (to be the devil`s advocate):

1. Knowing that we'll go with a simple saga after core v6 is released, would we still take this approach?
1. This would need to be backwards compatible, i.e. we'd need to query first to see if we can find a secondary index entity. If not, fallback to a full scan.

**Comment by Scooletz** [comment]
Yes, that's exactly how you described it! The cost of orphaned entities is not that big if any.

1. Could you link an issue that describes the simplification? As far as I understand, there's going to be only one key property allowed for saga. Is that right? Even with this, I'd still use this approach. I might not see the whole picture. If you see any gaps, please correct me.

1. Yes, and you described the introduction of this secondary indices in a perfect way! We can fallback to full scan, and if we do so, issue a task to update the secondary indices. The second query would hit the indices.

**Comment by mauroservienti** [comment]
should we consider this in scope for V6? if yes, could you please ad the expected label? cheers.

**Comment by yvesgoeleven** [comment]
Maintaining a secondary index can be done, but there are a few issues with it:

- Partition keys do not support all characters, so you cannot just take any property value and use it as an index (probably need to hash the value)
- Partition keys cannot change, so the content of property used for indexing cannot change from the initial value
- Every lookup now requires 2 queries
- Inserts and updates are no longer atomic and may lead to orphaned data
- Cleanup & Updates (MarkAsComplete) still requires a full table scan to find the related secondary indexes
- For existing customers we need a runtime migration strategy and stay backward compat

TBH I tried this about 2 years ago and eventually threw away the branch because it didn't work as great as I had hoped

**Comment by Scooletz** [comment]
@yvesgoeleven To address some of the concerns you've raised above:

- Yes, SHA1 + representing bytes in hex should work just fine
- That's true. I've thought that this is the rule ensured by core v6.
- Yes 
- Yes 
- That might be addressed by adding to the Saga entity additional list of all the keys with secondary indices. They instead of full table scan you have one get with PK+RK and n deletes PK+RK for each indexed value which can be issued concurrently/batched if any PK overlaps
- The description of migration strategy is provided above. When an index is missing, fallback to full scan + storying the index.

Could you write a bit more about what didn't work as great? Even if we drop this proposal, I'd love to hear more about that. The solution of storing indices PK+RK in saga entity removes the need for any scans for new sagas at all. In my opinion 2 lookups are better than one scan for high volume scenario. Unfortunately I don't know the real production distributions (how many concurrent sagas, how many properties) and haven't tested this solution. Looking forward to your opinion @yvesgoeleven

**Comment by SeanFeldman** [comment]
@Scooletz Simple Saga [proposal](https://github.com/Particular/PlatformDevelopment/issues/469) will eliminate need in this kind of complexity. It's just that we'd need to wait until v6 is out.

**Comment by SeanFeldman** [comment]
BTW, if not mistaken, storing list in storage tables as a property is not supported, only  primitive types. That could be solved by serializing the list, adding more complexity to an already complex solution.

**Comment by yvesgoeleven** [comment]
It was slow even with small data sets. Every operation is remote, and adds a minimum of 50 ms to your execution time. 
- You cannot batch the operations together because the PKey's differ. 
- As the operation is no longer atomic, the code to deal with idempotency also becomes more complex, adding even more time as you need to look up wether an operation has already been performed or not.
So every message handle performs at least 2 operations (Get, Persist), you basically double those and add some more to deal with idempotency. So you definitly feel it in terms of execution time. 

Another one of the main problems I had was picking the right property from the saga data to use as an index, you don't have any type metadata, nor the mapping info in the saga persister at the moment that you need it (Persist), so it is hard to choose. I solved this by searching for unique attribute using reflection and pick that property, but no guarantee that user will also use that property in saga mapping.

The perf loss on small sets, vs the potential in gains didn't add up imo

**Comment by Scooletz** [comment]
@SeanFeldman I've reread the proposal of SimpleSaga. Yes, having this on board will eliminate the whole need of secondary indices.
Yes, that's what's I meant. Using byte array as a bag for serialized list of PK+RK of index entries. Again that was meant to be a trade off between storage/complexity and speed.

**Comment by Scooletz** [comment]
@yvesgoeleven 
I think that you mentioned 50ms before, but I'll need to keep it my mind all the time. If that's the response time of a single storage call, that's a lot.
I agree with both points. I think that indices could be PUT (Insert Or Replace), but again you need to wait for storying them before saving saga, hence yes, response times add up as you described. We could go with a lookup (LRU cache for example) already persisted index entries or put even more layers on top of that, but this would become a really complex solution.
Summing up, seeing the SimpleSaga incoming, response times from Azure Storage Tables I see no further need of trying to solve the problem that soon will be solved by the core itself.

@SeanFeldman are you closing this issue then?

**Comment by yvesgoeleven** [comment]
@Scooletz it's roughly 50 ms per 1000km, it multiplies by distance, ask the aussies how much fun it is to use a service in US or europe (which are network layout wise nearer than asia btw)

**Comment by Scooletz** [comment]
@yvesgoeleven I'm really interested in geo-topic (Google Spanner, CockroachDB) and, again, will need to remind myself about that in cloud scenarios.

**Comment by Scooletz** [comment]
This is solved by the accepted PR #32
